src/classes/AuthenticationSession.ts

```diff
@@ -13,25 +13,94 @@ enum FusionAuthStatusCode {
 }
 
 export class AuthenticationSession {
-  public authToken = '';
+  private authToken = '';
+
+  public refreshToken = '';
 
   public readonly authContext;
 
+  private authTokenExpiresAt = new Date();
+
   private readonly fusionAuthClient;
 
   private twoFactorId = '';
 
   private twoFactorMethods: TwoFactorMethod[] = [];
 
-  public constructor(authContext: KeyboardAuthContext) {
+  private fusionAuthSftpAppId = '';
+
+  private fusionAuthSftpClientId = '';
+
+  private fusionAuthSftpClientSecret = '';
+
+  public constructor(
+    authContext: KeyboardAuthContext,
+    fusionAuthSftpAppId: string,
+    fusionAuthSftpClientId: string,
+    fusionAuthSftpClientSecret: string,
+  ) {
     this.authContext = authContext;
+    this.fusionAuthSftpAppId = fusionAuthSftpAppId;
+    this.fusionAuthSftpClientId = fusionAuthSftpClientId;
+    this.fusionAuthSftpClientSecret = fusionAuthSftpClientSecret;
     this.fusionAuthClient = getFusionAuthClient();
   }
 
   public invokeAuthenticationFlow(): void {
     this.promptForPassword();
   }
 
+  public async getToken() {
+    if (this.tokenWouldExpireSoon()) {
+      await this.getAuthTokenUsingRefreshToken();
+    }
+    return this.authToken;
+  }
+
+  private async getAuthTokenUsingRefreshToken(): Promise<void> {
+    try {
+      const clientResponse = await this.fusionAuthClient.exchangeRefreshTokenForAccessToken(
+        this.refreshToken,
+        this.fusionAuthSftpClientId,
+        this.fusionAuthSftpClientSecret,
+        '',
+        '',
+      );
+
+      if (clientResponse.response.access_token) {
+        this.authToken = clientResponse.response.access_token;
+        // The exchange refresh token for access token endpoint does not return a timestamp,
+        // it returns expires_in in seconds.
+        // So we need to create the timestamp to be consistent with what is first
+        // returned upon initial authentication
+        this.authTokenExpiresAt = new Date(
+          Date.now() + (clientResponse.response.expires_in ?? 1 * 1000),
+        );
+        logger.info('New access token obtained :', clientResponse.response);
+      } else {
+        logger.warn('No refresh token in response :', clientResponse.response);
+        this.authContext.reject();
```

Comment:
RELATED NOTE BUT PLEASE IGNORE IT: I'm now realizing that we've actually expanded the scope of `AuthenticationSession` to being BOTH the class that handles the establishment of an SFTP authentication session *and* being a class that manages FusionAuth tokens.  It is definitely conceptually confusing to have both in one place, and that leads to attempting to do things like this.

It might be best to have a new class called `AuthTokenManager`, which is solely responsible for managing the auth token / refresh token for the authentication session, contains the refresh token logic, and has no risk of accidentally using irrelevant (and dangerous) class attributes.

This would also be the only thing that has to be passed to the sftp session as well -- which would fix an original mistake of... why is an authSession passed to `SftpSessionHandler` in the first place!

That said, this feels out of scope.  I think what I'll do is actually just open an issue for this once the code is merged